File: revisiting_ctci/chapter3/StackMin_3_2.go
/*

 get the minimum value in a stack

 - efficient is to check on the way in and out
 - need to add a 'times seen' to account for the min getting popped 

*/ 

package main

import (
 	"./stack"
 	"fmt"
 	"errors"
)


func Min(s stack.Stack) (int , error) {
	if s.IsEmpty(){
		return -1, errors.New("The Stack is Empty!") 
	}
	min := s.Ord[0]
	for _, i := range s.Ord{
		if i < min{
			min = i
		}
	}

	return min, nil
}


func main(){

	test := stack.Stack{}
	test.Push(9)
	test.Push(2)
	test.Push(11)
	test.Push(7)
	test.Push(8)

	min, _ := Min(test)
	fmt.Println(min)
}

File: revisiting_ctci/chapter3/StackMin_3_2.py
"""
get the minimum value in a stack
"""

from stack import Stack


def MinStack(s):

	lowest = s.Data[0]
	
	for i in s.Data[1:]:
		if i < lowest:
			lowest = i
	return lowest


if __name__ == '__main__':

	test = Stack()
	for i in [9,3,1,4,7,5,5,]:
		test.Push(i)

	print(test)
	print('Minimum:\n')
	print(MinStack(test))		